cmd/dra-example-kubeletplugin/state.go

```diff
@@ -173,3 +206,97 @@ func (s *DeviceState) prepareDevices(claim *resourceapi.ResourceClaim) (Prepared
 func (s *DeviceState) unprepareDevices(claimUID string, devices PreparedDevices) error {
 	return nil
 }
+
+func (p *PreparedDevice) applyConfig() error {
+	return nil
+}
+
+// GetDevices extracts the list of drapbv1.Devices from PreparedDevices.
+func (pds PreparedDevices) GetDevices() []*drapbv1.Device {
+	var devices []*drapbv1.Device
+	for _, pd := range pds {
+		devices = append(devices, &pd.Device)
+	}
+	return devices
+}
+
+// ptrImplementsRuntimeObject is a helper interface for GetOpaqueDeviceConfig.
+//
+// It enforces that a pointer to the type parameter passed to
+// GetOpaqueDeviceConfig implements the runtime.Object interface as well as a
+// Validate() function.
+type ptrImplementsRuntimeObject[T any] interface {
+	*T
+	runtime.Object
+	Validate() error
+}
+
+// GetOpaqueDeviceConfig returns the config with the highest precedence specified for a device request in a resource claim.
+//
+// Configs can either be attached to the resource claim itself or to the
+// device class associated with the request. Configs found directly in the
+// resource claim take precedence over configs found in the device class.
+// Moreover, configs found earlier in the list of configs attached to either a
+// resource claim or a device class take precedence over configs found later in
+// their corresponding list.
+//
+// Only the config with the highest precedence in the order specified above will
+// be returned. All other configs will be ignored. If no config is found, the
+// default passed to this function will be returned instead.
+func GetOpaqueDeviceConfig[T any, TI ptrImplementsRuntimeObject[T]](
+	defaultConfig T,
+	decoder runtime.Decoder,
+	driverName string,
+	request string,
+	resourceClaim *resourceapi.ResourceClaim,
+	deviceClasses map[string]*resourceapi.DeviceClass,
+) (T, error) {
+	// Collect all configs in order of reverse precedence.
+	var allConfigs []resourceapi.DeviceConfiguration
+	for _, config := range deviceClasses[request].Spec.Config {
```

Comment:
This function should use the claim.status, not the claim.spec. It also shouldn't use DeviceClasses at all.

Reply:
Hmm. So are all the configs pre-merged into a single list in the claim status by the scheduler? Are they merged according to the semantics here (i.e. claim configs come before class configs so that their precedence is preserved)?

Reply:
Updated.